language/extensions/async/move-async-vm/tests/testsuite.rs

```diff
@@ -0,0 +1,403 @@
+    actor_metadata,
+    actor_metadata::ActorMetadata,
+    async_vm::{AsyncResult, AsyncSession, AsyncVM, Message},
+};
+use move_binary_format::access::ModuleAccess;
+use move_command_line_common::testing::EXP_EXT;
+use move_compiler::{
+    compiled_unit::CompiledUnit, diagnostics::report_diagnostics_to_buffer,
+    shared::NumericalAddress, Compiler, Flags,
+};
+use move_core_types::{
+    account_address::AccountAddress,
+    effects::ChangeSet,
+    identifier::{IdentStr, Identifier},
+    language_storage::{ModuleId, StructTag},
+    resolver::{ModuleResolver, ResourceResolver},
+};
+use move_prover_test_utils::{baseline_test::verify_or_update_baseline, extract_test_directives};
+use move_vm_types::gas_schedule::GasStatus;
+use std::{
+    cell::RefCell,
+    collections::{BTreeMap, BTreeSet, VecDeque},
+    path::{Path, PathBuf},
+    str::FromStr,
+};
+
+const TEST_ADDR: &str = "0x3";
+const SOURCE_DIRS: &[&str] = &[
+    "./tests/sources",
+    "../move-async-lib/sources",
+    "../../../move-stdlib/sources",
+];
+
+struct Harness {
+    module_cache: BTreeMap<Identifier, CompiledUnit>,
+    vm: AsyncVM,
+    actor_instances: Vec<(ModuleId, AccountAddress)>,
+    baseline: RefCell<String>,
+    resource_store: RefCell<BTreeMap<(AccountAddress, StructTag), Vec<u8>>>,
+}
+
+fn test_account() -> AccountAddress {
+    AccountAddress::from_hex_literal(TEST_ADDR).expect("valid test address")
+}
+
+fn test_runner(path: &Path) -> datatest_stable::Result<()> {
+    let target_module = path
+        .with_extension("")
+        .file_name()
+        .unwrap()
+        .to_string_lossy()
+        .to_string();
+    let deps = extract_test_directives(path, "// dep:")?;
+    let actors = extract_test_directives(path, "// actor: ")?;
+    let instances = extract_test_directives(path, "// instance: ")?;
+    let harness = Harness::new(
+        std::iter::once(target_module.clone())
+            .chain(deps.into_iter())
+            .collect(),
+        actors,
+        instances,
+    )?;
+    harness.run(&target_module)?;
+    let baseline_path = path.with_extension(EXP_EXT);
+    verify_or_update_baseline(baseline_path.as_path(), harness.baseline.borrow().as_str())?;
+    Ok(())
+}
+
+datatest_stable::harness!(test_runner, "tests/sources", r".*\.move");
+
+// ========================================================================================
+// Test execution
+
+impl Harness {
+    fn run(&self, _module: &str) -> anyhow::Result<()> {
+        let mut gas = GasStatus::new_unmetered();
+        let mut tick = 0;
+        // Publish modules.
+        let mut proxy = HarnessProxy { harness: self };
+        let mut session = self.vm.new_session(test_account(), 0, &mut proxy);
+        let mut done = BTreeSet::new();
+        for id in self.module_cache.keys() {
+            self.publish_module(&mut session, id, &mut gas, &mut done)?;
+        }
+        // Initialize actors
+        let mut mailbox: VecDeque<Message> = Default::default();
```

Comment:
It looks like the "init" command does not go through the mailbox. Perhaps in Async Move, you don't think "init" as a command. In Jazz, at the moment, there is nothing special about the init command. Any command could be put into the starting position of the log and executed as the first command.  Just thought I would mention it.

Reply:
The difference might be that in Move, you cannot have a struct with default values initialized. So you can't have `#[message] fun init(this: &mut Actor)`. On the other hand, the inits are restricted and can't send any messages. So to simulate the Jazz model, one would have `init` just set defaults in the state, and then some `#[message] fun start.`